Shell/AST.h

```diff
@@ -0,0 +1,730 @@
+/*
+ * Copyright (c) 2020, the SerenityOS developers.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include "Forward.h"
+#include "Job.h"
+#include <AK/RefCounted.h>
+#include <AK/RefPtr.h>
+#include <AK/String.h>
+#include <AK/Types.h>
+#include <AK/Vector.h>
+#include <LibLine/Editor.h>
+
+using TheExecutionInputType = RefPtr<Shell>;
+
+namespace AST {
+
+struct HighlightMetadata {
+    bool is_first_in_list { true };
+};
+
+struct Position {
+    size_t start_offset { 0 };
+    size_t end_offset { 0 };
+    bool in_range(size_t offset) const { return start_offset <= offset && end_offset >= offset; }
+};
+
+struct Redirection {
+    String path;
+    int fd { 0 };
+    enum {
+        Read,
+        Write,
+        WriteAppend,
+    } direction { Read };
+};
+
+struct Rewiring {
+    int source_fd { 0 };
+    int dest_fd { 0 };
+    enum class Close {
+        None,
+        Source,
+        Destination,
+    } must_be_closed { Close::None };
+};
+
+struct Command {
+    Vector<String> argv;
+    Vector<Redirection> redirections;
+    Vector<Rewiring> rewirings;
```

Comment:
What definitely matters is which order redirections & rewirings are applied in. E.g. in Bash:

```
$ ls -l /proc/self/fd 9>/tmp/foo 8>&9 9>/tmp/bar
total 0
lrwx------. 1 sergey sergey 64 Jun 17 18:20 0 -> /dev/pts/2
lrwx------. 1 sergey sergey 64 Jun 17 18:20 1 -> /dev/pts/2
lrwx------. 1 sergey sergey 64 Jun 17 18:20 2 -> /dev/pts/2
lr-x------. 1 sergey sergey 64 Jun 17 18:20 3 -> /proc/2236655/fd
l-wx------. 1 sergey sergey 64 Jun 17 18:20 8 -> /tmp/foo
l-wx------. 1 sergey sergey 64 Jun 17 18:20 9 -> /tmp/bar
```

So you cannot just keep them in two separate vectors.

Reply:
hmmm...then don't do it that way?
I always found the order of those rather confusing, file redirections first, other redirections next feels more natural...but I guess I can try preserving their order.
The separate vectors _will_ stay.

Reply:
Feel free to keep the vectors separate if you can make it work as intended — though I'd make redirection into an abstract class or something that has a method to "apply itself".

> I always found the order of those rather confusing, file redirections first, other redirections next feels more natural.

Umm no? They're executed in order, much like the equivalent C code:

```c
open_as("/tmp/foo", 9);
dup2(9, 8);
open_as("/tmp/bar", 9);

void open_as(const char* path, int requested_fd) {
    int opened_fd = open(path, O_RDONLY);
    // ...handle errors...
    dup2(opened_fd, requested_fd);
    if (opened_fd != requested_fd)
        close(opened_fd);
}
```

You wouldn't suggest a C compiler should reorder dup and open calls because it feels more natural, would you? :slightly_smiling_face:

Reply:
I would tbh lol.
but sure, we'll get there.